packages/react-reconciler/src/__tests__/ReactIncremental-test.internal.js

```diff
@@ -1421,7 +1422,7 @@ describe('ReactIncremental', () => {
     ]);
   });
 
-  it('does not call static getDerivedStateFromProps for state-only updates', () => {
+  it('calls getDerivedStateFromProps even for state-only updates', () => {
```

Comment:
Why this change? Do I need to update my diagram? 😬

Does this fix some edge case?

Reply:
`getDerivedStateFromProps` is inherently weird because it treats receiving props as an event, which is not how React works. With this change, we call it on every render, not only on prop changes, and we only persist the result to the queue once there's no more work to be done. So the result still accumulates, but it's more deterministic — not completely deterministic, but it works for most cases. The implementation is also simpler, because we don't have to enqueue an update.

If you're use `getDerivedStateFromProps` the way it's intended to be used, then this change shouldn't affect anything. If you're using it to count the number of renders, or to accumulate prop change "events", then this is breaking. We (Brian, Sebastian, and I) decided it was important enough to put in a minor. Call it a bugfix.

Reply:
The three use cases for `getDerivedStateFromProps` are:

- Memoization. We have a better model for how this should work in the future.
- Avoiding the need to derive the same value repeatedly in multiple methods/lifecycles. Again, our future model handles this better.
- Reacting to prop changes. This is the weird one, and I don't think we totally understand yet what the best practices should be.

Reply:
If we back out of this change let's also remember to back out of https://github.com/facebook/react/pull/12676.

Reply:
Going from React 16.3 to 16.4 has caused two major regressions in my application because of this change. I was relying on the fact that getDerivedStateFromProps() was only called when the props changed, in the same way componentWillReceiveProps worked.

Granted, I probably did not understand the original intent of getDerivedStateFromProps. My case is this is more than just a simple bug fix.

Reply:
@guillaumep I'm sorry to hear that 16.4 caused troubles for you.

The point you raise is discussed in detail in #12898 if you'd like to learn more about why we made it.